### PR TITLE
Add primitive compiletime operations on singleton types

### PR DESCRIPTION
**Pull request by MaximeKjaer**
This PR adds support for type-level operations on singletons, making it possible to write the following:

```scala
val x: 2 + 3 = 5
val y: 3 * 4 = 12
```

Having these primitive operations makes it possible to define more complex operations through match types, e.g.:

```scala
type GCD[A <: Int, B <: Int] <: Int = B match {
    case 0 => A
    case _ => GCD[B, A % B]
}
val z: GCD[252, 105] = 21
```

## Background
With `scala.compiletime.S` and match types, it's possible to implement a variety of operations on singletons. For instance, addition and multiplication can be implemented as follows:

```scala
type +[A <: Int, B <: Int] <: Int = A match {
    case 0 => B
    case S[aMinusOne] => S[aMinusOne + B]
}

type *[A <: Int, B <: Int] <: Int = A match {
    case 0 => 0
    case _ => MultiplyLoop[A, B, 0]
}

type MultiplyLoop[A <: Int, B <: Int, Acc <: Int] <: Int = A match {
    case 0 => Acc
    case S[aMinusOne] => MultiplyLoop[aMinusOne, B, B + Acc]
}
```

However, as these are implemented on the low-level primitive of `scala.compiletime.S`, they are highly recursive. Interpreting these match types quickly runs out of memory (for instance, `100 * 100` exceeds the recursion limit on default settings).

## Further work and discussion
A natural extension to this would be to be able to type term-level operations as the precise type-level equivalent, perhaps as an opt-in, like `val x: 1 + 2 = 1 + 2`.

## Related
- @propensive actually [already had an implemention](https://github.com/propensive/dotty/commit/e616c57324df417b8d273352e15ea46d3eca993a) of addition and multiplication on singleton ints, as [pointed out here](https://github.com/lampepfl/dotty/issues/7512#issuecomment-551015857).
- The [singleton-ops](https://github.com/fthomas/singleton-ops) library provides similar functionality for Scala 2.X.
- Dotty implements operator precedence for infix type aliases, so these operation types compose quite nicely.

### CODE REVIEW

**Comment by soronpo** [comment]
This is exciting! I contributed a lot to singleton-ops, and I wouldn't mind at all to retire it in favor of true compiler support. I will provide extensive feedback when I'll review the work.

**Comment by soronpo** [comment]
Topics to discussion:
1. What about `Long`, `Char`, `Float`, `Double`, and `String` operations that singleton-ops supports? 
2. If say we add `Long` operations support, what do we do for a `Long` + `Int` singleton operation?
3. What do you think about compile-time `NumberOfLeadingZeros` (`Java.lang.Integer.numberOfLeadingZeros`). I use it from singleton-ops.

**Comment by MaximeKjaer** [comment]
It seems that I've broken some tests. I'll work on fixing those.

@soronpo, thank you! I think the guiding principle should be to aim for support for all primitive operations on all supported singleton types. It would seem arbitrary to decide that certain types or certain operations should not be supported, especially when adding support for a primitive operation is as trivial as it is.

So to answer your points:

1. Those should also be supported. For now, I'm thinking of splitting the work in two: first this PR as a proof of concept, and if it's merged, I'll make a second PR aiming for completeness.
2. The standard library does this by overloading the `+` method, but overloading a type alias doesn't work in Scala:
	
	```scala
	type +[X <: Int, Y <: Int] <: Int
	type +[X <: Int, Y <: Float] <: Float // Error: + is already defined as type +
	```
	
	However, it may possible to emulate overloading by using union types and match types, as follows:

	```scala
	type That = Double | Float | Long | Int | Char | Short | Byte | String
	type Result[X <: That] = X match {
		case Double => Double
		case Float => Float
		case Long => Long
		case String => String
		case _ => Int
	}

	type +[X <: Int, Y <: That] <: Result[Y]
	```

	With this solution, the types indicate something slightly more general than what will be supported by the compiler. For instance, `1 + (2 | 3L)` is not reduced, but the type checker will not complain about this type application.

	This example opens up a discussion on how to deal with union types. I think it could make sense to distribute over the union, so that the resulting type would be `3 | 4L`. This is not something I have implemented yet, but it's an interesting discussion.
	
3. I think that should be supported too.

File: library/src/scala/compiletime/int/package.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package scala.compiletime
+
+package object int {
+  type +[X <: Int, Y <: Int] <: Int
+  type -[X <: Int, Y <: Int] <: Int
+  type *[X <: Int, Y <: Int] <: Int
+  type /[X <: Int, Y <: Int] <: Int
+  type %[X <: Int, Y <: Int] <: Int
+
```
[the remark this answers is not in the document]
Comment (neko-kai) [reply library/src/scala/compiletime/int/package.scala]: `String` is `AnyRef`, so it'll have to be just `<: Any` (aka no bound)

[the remark this answers is not in the document]
Comment (MaximeKjaer) [reply library/src/scala/compiletime/int/package.scala]: That's a good idea, I like the name `scala.compiletime.ops`. I've made that change in cb3ab697b8443d5d8de687e244ec935d195b43ba.

As for the bounds, I agree. I think that change is best made in the PR that will introduce support for all singleton types, so I'll leave it as is for now, but refer back to this discussion in the next PR.

File: tests/neg/compiletime-singleton-ops.scala [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import scala.compiletime._
+import scala.compiletime.int._
+import scala.compiletime.boolean._
+
+object Test {
+  val t0: 2 + 3 = 5
```
[the remark this answers is not in the document]
Comment (MaximeKjaer) [reply tests/neg/compiletime-singleton-ops.scala]: This is a good point, thank you! I've added such tests in b458372b3705b3cff56c5b46380830a391230bcd.

File: tests/neg/compiletime-singleton-ops.scala [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  val t1: 2 + 2 = 5 // error
+  val t2: -1 + 1 = 0
+  val t3: -5 + -5 = -11 // error
+
+  val t4: 10 * 20 = 200
```
[the remark this answers is not in the document]
Comment (MaximeKjaer) [reply tests/neg/compiletime-singleton-ops.scala]: Yes, they do work! Operator priority even works out of the box, which is nice. I've added tests in b458372b3705b3cff56c5b46380830a391230bcd, in `tests/pos/singleton-ops-composition.scala`.

File: library/src/scala/compiletime/int/package.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaximeKjaer) [reply library/src/scala/compiletime/int/package.scala]: Having moved everything to `scala.compiletime.ops`, it would make sense to also move `scala.compiletime.S` to `scala.compiletime.ops.S`. However, since that would be a breaking change, I think it's best to keep that for a separate PR and have that discussion there.

**Comment by soronpo** [comment]
What do you think about an `type a = Int + 1` (meaning, a non-singleton addition with a singleton)?
In the singleton-ops I created something called `TwoFace` values, which is a mechanism to allow best-effort inlining. Example:
```scala
val oneCT : TwoFace.Int[1] = TwoFace.Int(1)
val one : Int = 1
val oneRT :  TwoFace.Int[Int] = TwoFace.Int(one)

val twoCT : TwoFace.Int[2] = oneCT + oneCT
val twoRT1 : TwoFace.Int[Int] = oneCT + oneRT
val twoRT2 : TwoFace.Int[Int] = oneRT + oneCT
val twoRT3 : TwoFace.Int[Int] = oneRT + oneRT
```

Currently the dotty inline parameters mechanism only allows compile-time values, so I'm still forced to use the TwoFace concept. So for this to work, I require `Int + 1 =:= Int`. If it's not an acceptable default behavior then perhaps we can have a separate `flexible_ops` (or a better named) package.

**Comment by soronpo** [comment]
To clarify further, ideally I would like to be able to write:
```scala
object TwoFace {
  opaque type Int[T <: scala.Int] = scala.Int
  object Int {
    def apply[T <: scala.Int](value : T) : Int[T] = value
    protected def force[T <: scala.Int](value : scala.Int) : Int[T] = value
    def [L <: scala.Int, R <: scala.Int](left : Int[L]) + (right : Int[R]) : Int[L + R] = force[L + R](left + right) 
  }
}
```

**Comment by MaximeKjaer** [comment]
This is a good point, thank you! I think typing `Int + 1` as an `Int` is acceptable. It's consistent with `S[Int]` (for which `val x: S[Int] = 1` type-checks), and it allows for more graceful failure when non-singleton values are provided, like in your example of addition on `Int`.

I've implemented it in 4d73d89131d05b33df285e97b34241f838c212a8 by falling back to a type check with the super-type when constant folding fails.

<details>
<summary>With this change, I'm able to make the program below compile</summary>

```scala
import scala.compiletime.ops._

object TwoFace {
  opaque type Int[T <: scala.Int] = scala.Int
  object Int {
    def apply[T <: scala.Int](value : T) : Int[T] = value
    protected def force[T <: scala.Int](value : scala.Int) : Int[T] = value
    def [L <: scala.Int, R <: scala.Int](left : Int[L]) + (right : Int[R]) : Int[L + R] = force[L + R](left + right)
  }
}

object TestTwoFace {
  import TwoFace.Int.+

  val oneCT : TwoFace.Int[1] = TwoFace.Int(1)
  val one : Int = 1
  val oneRT :  TwoFace.Int[Int] = TwoFace.Int(one)

  val twoCT : TwoFace.Int[2] = oneCT + oneCT
  val twoRT1 : TwoFace.Int[Int] = oneCT + oneRT
  val twoRT2 : TwoFace.Int[Int] = oneRT + oneCT
  val twoRT3 : TwoFace.Int[Int] = oneRT + oneRT
}
```
</details>

**Comment by soronpo** [comment]
This is awesome! We need to make sure we don't screw up invariance.
So:
```scala
val invarianceError : TwoFace.Int[Int] = TwoFace.Int[1](1) //error
val worksDueToWidening : TwoFace.Int[Int] = TwoFace.Int(1) //works due to widening
val weirdButOK : TwoFace.Int[Int+0] = TwoFace.Int[1](1)
```
<details>
  <summary>Of course, I would change the `TwoFace` implementation above to be covariant:</summary>
  
```scala
import scala.compiletime.ops._

object TwoFace {
  opaque type Int[+T <: scala.Int] = scala.Int
  object Int {
    def apply[T <: scala.Int](value : T) : Int[T] = value
    protected def force[T <: scala.Int](value : scala.Int) : Int[T] = value
    def [L <: scala.Int, R <: scala.Int](left : Int[L]) + (right : Int[R]) : Int[L + R] = force[L + R](left + right)
  }
}

object TestTwoFace {
  import TwoFace.Int.+

  val oneCT : TwoFace.Int[1] = TwoFace.Int(1)
  val one : Int = 1
  val oneRT :  TwoFace.Int[Int] = TwoFace.Int(one)
  val oneRT2 :  TwoFace.Int[Int] = TwoFace.Int[1](1) //works due to covariance

  val twoCT : TwoFace.Int[2] = oneCT + oneCT
  val twoRT1 : TwoFace.Int[Int] = oneCT + oneRT
  val twoRT2 : TwoFace.Int[Int] = oneRT + oneCT
  val twoRT3 : TwoFace.Int[Int] = oneRT + oneRT
}
```
</details>

**Comment by soronpo** [comment]
Maybe now it will be easier to convince the SIP committee to support unary types so we can replace `Negate[T]` and `![T]` with normal looking code :smile:

**Comment by soronpo** [comment]
This might be a little more tough, but what about implementing algebraic rules within the type system (for the discussed ops)?
```scala
inline def commutative[A <: Int, B <: Int](implicit ev : A + B =:= B + A) : Unit = {}
inline def associative[A <: Int, B <: Int, C <: Int](implicit ev : A + (B + C) =:= (A + B) + C) : Unit = {}
//others rules...
```
Of course, in case of overflows or precision issues, these rules may not apply, so I'm open to other opinions on the matter.

File: library/src/scala/compiletime/int/package.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (soronpo) [reply library/src/scala/compiletime/int/package.scala]: After thinking about it a little more, I think you're right and we need to set the upper bound using match type, otherwise it will not be possible to use the same operation for different types and place the result where there is an expected upper-bound.

**Comment by MaximeKjaer** [comment]
The covariant implementation you gave pretty much works as expected, so that's good news. The only line that I cannot compile in your example is the following:

```scala
val twoCT : TwoFace.Int[2] = oneCT + oneCT
```

I've narrowed down the problem to type inference (and not a bug in the `scala.compiletime.ops`implementation, thankfully!). Running the compiler with `-Xprint:frontend` tells us that the type parameters of `TwoFace.Int.+` are inferred as follows:

```scala
val twoCT: TwoFace.Int[2.type] = 
      TwoFace.Int.+[Int, Int](TestTwoFace.oneCT)(TestTwoFace.oneCT)
```

I'm not exactly sure of why it infers `Int` instead of `1.type`, especially seeing that `oneCT` is correctly typed as `TwoFace.Int[1.type]`. It seems like the type inference is being a little too conservative in this instance. There may be a good reason for this, so I will try to find out; if there is no good reason for it, I can open an issue or submit a PR.

Helping out the type inference by explicitly writing the type parameters solves the problem, and makes the line compile:

```scala
val twoCT: TwoFace.Int[2] = TwoFace.Int.+[1, 1](oneCT)(oneCT)
```

**Comment by MaximeKjaer** [comment]
What would be a good use-case for these algebraic rules?

As a small aside, the examples you gave (commutativity and associativity) behave more like a test, rather than like a law that can be proven. Because these rules are true (for `Int` at least), they will always be able to find an implicit. However, I could also write a rule that is incorrect in general, but happens to be true when `B` has type `0`:

```scala
def badLaw[A <: Int, B <: Int](implicit ev: A + B =:= A - B) : Unit = {}
```

Because `Int + Int =:= Int`, the rule also passes with `Int` type parameters, which may be a little counter-intuitive at first. All the following lines will compile:

```scala
badLaw[Int, Int]
badLaw[Int, 20]
badLaw[1, 0]
badLaw[0, 0]
```

It only complains if I can give it a counter-example:

```scala
badLaw[3, 1] // error, implicit not found
```

**Comment by soronpo** [comment]
> What would be a good use-case for these algebraic rules?
Not much of a use-case, but the goal is regularity:

```scala
import scala.compiletime.ops._

object TwoFace {
  opaque type Int[T <: scala.Int] = scala.Int
  object Int {
    def apply[T <: scala.Int](value : T) : Int[T] = value
    protected def force[T <: scala.Int](value : scala.Int) : Int[T] = value
    def [L <: scala.Int, R <: scala.Int](left : Int[L]) + (right : Int[R]) : Int[L + R] = force[R + L](left + right) //notice the switched positions
  }
}
```

**Comment by OlivierBlanvillain** [comment]
I'm not convinced about `Int * 2 =:= Int`, these two types do not denote the same set of values... Don't we want subtyping in one direction instead?

**Comment by soronpo** [comment]
> I'm not convinced about `Int * 2 =:= Int`, these two types do not denote the same set of values... Don't we want subtyping in one direction instead?

It won't work with one direction. If `Int * 2 <:< Int` and `Int * 3 <:< Int`, how `Int * 2` and `Int * 3` are related? 
The point is that if you're having any operation between a singleton type and its super type you will always get the super type. I don't see the problem here.

**Comment by soronpo** [comment]
To clarify, this is exactly what happens at the term level:
```scala
val one : Int = 1
one * 2  // : Int
one * 3 // : Int
1 * 2 // : 2
1 * 3 // : 3
```

**Comment by smarter** [comment]
> It won't work with one direction. If Int * 2 <:< Int and Int * 3 <:< Int, how Int * 2 and Int * 3 are related?

What do you mean by "won't work" ? you can have two subtypes of a given type whose sets of values partially overlap.

**Comment by soronpo** [comment]
> What do you mean by "won't work" ? you can have two subtypes of a given type whose sets of values partially overlap.

Well, this actually depends on whether or not algebraic rules are applied to the typesystem.
```scala
val one : Int = 1
val tfA = TwoFace.Int(one) + TwoFace.Int(1)
val tfB = TwoFace.Int(1) + TwoFace.Int(one)
summon[tfA.type =:= tfB.type] //error
```
This error is weird. 
I guess subtyping might be OK, but it just aggravates the lack of algebraic rules.

**Comment by smarter** [comment]
When `A <:< B` is false, it never means "A is not a subtype of B", it means "I could not prove that A is a subtype of B".

**Comment by OlivierBlanvillain** [comment]
I think adding algebraic rules on integer operations is outside the scope of this PR. The first step, and what's the most urgently missing at the moment, is a way *perform* integer operations on singleton types.

One argument to limit the scope to only reduce when both types are singleton would be that doing something more elaborate might embed more knowledge in the compiler than there is in the runtime. The JVM know about `1 + 2 = 3`, but has no idea about `A + B = B + A`, so let's first get on par with the runtime before considering going further...

**Comment by MaximeKjaer** [comment]
I think that @OlivierBlanvillain raises a very good point. Currently, the following type-checks:

```scala
object TwoFace {
  opaque type Int[T <: scala.Int] = scala.Int
  object Int {
    def apply[T <: scala.Int](value : T) : Int[T] = value
    protected def force[T <: scala.Int](value : scala.Int) : Int[T] = value
    def [L <: scala.Int, R <: scala.Int](left : Int[L]) / (right : Int[R]) : Int[L / R] = force[R / L](left + right) // switched positions
  }
}
```

This may be fine for addition, which is commutative, but it's not acceptable for division, in my opinion.

I suggest that we go back to only reducing singleton args for now, and not cases like `Int + 1`. I think the discussion we've already had shows that the latter should maybe be tied to slightly more complex machinery (like support for algebra), which I think is out of the scope of this PR. It's good that we've had some discussion about it now already, so that we know what's ahead of us, but I suggest that we take it up again in another PR more specifically focused on the topic.

So to sum up the conversation thus far:

## Remains in this PR
1. Removing handling of cases like `Int + 1`
2. Adding an error when dividing by 0 (or modulo by 0)

Once I've done those two things, I think the PR is ready to be merged.

## Possible extensions, for future PRs
1. Completeness: support for `Long`, `Float`, etc (as discussed [here](https://github.com/lampepfl/dotty/pull/7628#issuecomment-559079755))
2. Algebraic rules (as discussed [here](https://github.com/lampepfl/dotty/pull/7628#issuecomment-560558852))
3. Unary types and SIP-36 (as discussed [here](https://github.com/lampepfl/dotty/pull/7628#issuecomment-560480824))
4. More precise type inference (as discussed [here](https://github.com/lampepfl/dotty/pull/7628#issuecomment-561573448))

**Comment by soronpo** [comment]
>I suggest that we go back to only reducing singleton args for now, and not cases like Int + 1. 

I'm inclined to agree because covariance should handle this properly. 
If it's `TwoFace.Int[+T]`, then this should still work:
```scala
val one : Int = 1
val a : TwoFace.Int[Int] = TwoFace.Int(one) + TwoFace.Int(1) 
```

> So to sum up the conversation thus far:

One important thing- the covariance error you mention [here](https://github.com/lampepfl/dotty/pull/7628#issuecomment-561573448) must be dealt with, IMO.

**Comment by soronpo** [comment]
For proof of greater usability, I think this PR should at least add the following:
* `String + String` concatenation.
* `ToString[T]` conversion from `Int`.
* `Require[Cond <: Boolean, Msg <: String]` to create logic constraints on values (if `Cond` is untrue generate the error message `Msg`).
All the above will bring us much further for people to be able to use this feature to replace the libraries `refined` and `singleton-ops`.

Example code:
```scala
import scala.compiletime.ops._

opaque type Positive = Int
object Positive {
  type Check[T] = Require[T > 0, "The value provided isn't positive. Found: " + ToString[T]]
  def apply[T <: Int](value : T)(given Check[T]) : Positive = value
}

val eight = Positive(8)
val negError = Positive(-1) //error: The value provided isn't positive. Found: -1
```

**Comment by soronpo** [comment]
Instead of `Require`, we can just have `Error[Msg]` special case in the compiler. We can then write:
```scala
type Require[Cond <: Boolean, Msg <: String] = Cond match {
  case true => Nothing
  case false => Error[Msg]
}
```

**Comment by soronpo** [comment]
> All the above will bring us much further for people to be able to use this feature to replace the libraries `refined` and `singleton-ops`.

Oh, it seems I forgot about inline methods :man_facepalming: 
Then I don't think there is need for this.

**Comment by soronpo** [comment]
> > All the above will bring us much further for people to be able to use this feature to replace the libraries `refined` and `singleton-ops`.
> 
> Oh, it seems I forgot about inline methods man_facepalming
> Then I don't think there is need for this.

Well, it turns out that it is currently impossible to mix opaque types and inline defs, so using the compiletime ops for this will be nice. I created an issue for the inline defs (feature-lacking) problem: https://github.com/lampepfl/dotty-feature-requests/issues/82

**Comment by MaximeKjaer** [comment]
> The most important thing- the covariance error you mention here must be dealt with, IMO.

I agree, and I think we can open an issue for now. As far as I can tell, it's a type inference bug, not a bug with this implementation or with covariance, so I'm inclined to treat them separately. If merged, this PR should provide some nice motivation to fix it, IMO.

As for the other points:

- I haven't added `String + String` because that would require overloading the `+` type. As we discussed earlier, I think that merits its own PR.
- I've added `ToString[T <: Int]`
- I haven't added a `Require` or an `Error` on this branch just yet. I am more in favor of an `Error[Msg <: String]` that mirrors `scala.compiletime.error`, and I have a [draft implementation of that](https://github.com/MaximeKjaer/dotty/commit/a6ed21e9b114cd9c17b7bc70288c3c902a5118f8) that generates the nice error below on [this program](https://github.com/MaximeKjaer/dotty/commit/a6ed21e9b114cd9c17b7bc70288c3c902a5118f8#diff-4d2cdb7e57c07ca9fda2d666253566f1). I think requires a bit more thorough testing, for now I've just hacked it together.
   
  <details><summary>Generated Error</summary>
  
  ```scala
  -- Error: tests/playground/Errors.scala:22:32 ----------------------------------
  22 |  val y: Positive = Positive[-1](-1)
     |                                ^
     |                                The value provided isn't positive
  one error found
  ```
  </details>

Preferably, I'd like to keep changes somewhat incremental rather than have a big PR doing many different things. So I think this is ready for your review @OlivierBlanvillain, and if it's all fine and the tests pass, I think we can merge.

**Comment by soronpo** [comment]
OK, I think it's better to have an `.md` documentation.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 4]
```diff
@@ -3595,19 +3595,79 @@ object Types {
...
-          }
-        else tryMatchAlias
+
+        tryCompiletimeConstantFold.getOrElse(tryMatchAlias)
+
       case _ =>
         NoType
     }
 
+    def tryCompiletimeConstantFold(implicit ctx: Context): Option[Type] = tycon match {
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: No need to return an `Option` here, you can just return a `Type` and use `NoType` instead of `None`. (Types have an `orElse` method that would replace the `getOrElse` in `tryNormalize`)

File: library/src/scala/compiletime/ops/package.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package scala.compiletime
+
+import scala.annotation.infix
+
+package object ops {
+  @infix type ==[X <: AnyVal, Y <: AnyVal] <: Boolean
+  @infix type !=[X <: AnyVal, Y <: AnyVal] <: Boolean
+
+  @infix type +[X <: Int, Y <: Int] <: Int
```
Comment (OlivierBlanvillain) [inline library/src/scala/compiletime/ops/package.scala]: These ops need prefixes, otherwise it's going to be hard to add `+[S <: String, T <: String]`...

Comment (OlivierBlanvillain) [reply library/src/scala/compiletime/ops/package.scala]: Maybe something like this?:

```
package object ops {
  type ==
  type !=
  
  object int {
    ...
  }
   
  object boolean {
    ...
  }
}
```

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 6]
```diff
@@ -898,6 +899,20 @@ class Definitions {
   final def isCompiletime_S(sym: Symbol)(implicit ctx: Context): Boolean =
     sym.name == tpnme.S && sym.owner == CompiletimePackageObject.moduleClass
 
+  final def isCompiletimeAppliedType(sym: Symbol)(implicit ctx: Context): Boolean = {
+    def isOpsPackageObjectAppliedType: Boolean =
+      sym.owner == CompiletimeOpsPackageObject.moduleClass && Set(
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: I'll put that `Set` in a val to avoid recomputing it on every call

File: library/src/scala/compiletime/ops/package.scala (hunk 5 above)
Comment (soronpo) [reply library/src/scala/compiletime/ops/package.scala]: I don't agree. When you write A + B you don't need to think what to import. Same for terms, same for types. In future PR all these operation types should be multipurpose (accept several different supertype combinations). 
If it's better to do `<: AnyVal` or `<: Boolean | Int` I don't know.

Comment (OlivierBlanvillain) [reply library/src/scala/compiletime/ops/package.scala]: The issue is that `A + B` means `+[A, B]` and not `A#+[B]`, so we need a way to resolve that `+`...

The goal is to have type equivalants for term operations, and pretending that there is a `+` defined on `AnyVal` doesn't really go into that direction.

But it should be possible to define your own `+`:

```scala
type +[A <: AnyVal, B <: AnyVal] =
  (A, B) match {
    case (Int, Int)       => A obs.int.+ B
    case (String, String) => A obs.string.+ B
    case _ => Error["type error"]
  }
```

Comment (soronpo) [reply library/src/scala/compiletime/ops/package.scala]: So you end up with the same thing. I doubt it's a lot better than handling the supported combinations directly in the compiler under the same symbol.

Comment (MaximeKjaer) [reply library/src/scala/compiletime/ops/package.scala]: The problem with defining `+` both in `ops.int` and `ops.string` is that this seemingly innocent code actually is a compilation error:

```scala
import scala.compiletime.ops.int._
import scala.compiletime.ops.string._

summon[1 + 2]
//       ^ Reference to + is ambiguous
//         it is both imported by import scala.compiletime.ops.int._
//         and imported subsequently by import scala.compiletime.ops.string._
```

An additional problem is that qualified types cannot be used infix:

```scala
import scala.compiletime.ops._
summon[1 int.+ 2]
//          ^ an identifier expected, but '.' found
```

I agree with @OlivierBlanvillain that prefixing overloaded ops helps with the implementation -- it makes it simpler to constant fold if the name analysis phase already has determined which interpretation of `+` we're talking about. However, in the light of the above issues, I also agree with @soronpo that unprefixed ops make for a better user experience. I've managed to find a middle ground in 7b9ccb41aa0807a95168278dc254127feeb71a01, with a match type redirecting to the correct prefixed version, as Olivier suggested.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 6 above)
Comment (MaximeKjaer) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: Good call. Done in 5e287cda8b18dd9c90bff69eb782e23afb820af9

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 4 above)
Comment (MaximeKjaer) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: :+1: Done in 5e287cda8b18dd9c90bff69eb782e23afb820af9

File: library/src/scala/compiletime/ops/package.scala [hunk 7]
```diff
@@ -6,39 +6,33 @@ package object ops {
   @infix type ==[X <: AnyVal, Y <: AnyVal] <: Boolean
   @infix type !=[X <: AnyVal, Y <: AnyVal] <: Boolean
 
-  @infix type +[X <: Int | String, Y <: Int | String] = (X, Y) match {
```
Comment (soronpo) [inline library/src/scala/compiletime/ops/package.scala]: I actually like this example a lot. It's a shame to delete it. I propose to change it to a test code (or provide it as example in a document).

File: library/src/scala/compiletime/ops/package.scala [hunk 8]
```diff
@@ -6,39 +6,33 @@ package object ops {
   @infix type ==[X <: AnyVal, Y <: AnyVal] <: Boolean
```
Comment (soronpo) [inline library/src/scala/compiletime/ops/package.scala]: I believe `==` and `!=` should be split as well. Either everything is split according to the supertype or non of it is.

File: library/src/scala/compiletime/ops/package.scala (hunk 7 above)
Comment (MaximeKjaer) [reply library/src/scala/compiletime/ops/package.scala]: Good point -- I've done both in 2cab5919559cca0836d517b3fc74475d3336fe4c, and generally documented the feature.

File: library/src/scala/compiletime/ops/package.scala (hunk 8 above)
Comment (MaximeKjaer) [reply library/src/scala/compiletime/ops/package.scala]: I moved them into `scala.compiletime.ops.any` in 8c117c1342913a8d23cf59cd2321afbbecc7e952.

The alternative was to duplicate them into each subpackage for each supported type, and then duplicate the constant folding code... which felt like a lot of duplication. Seeing that `==` and `!=` are [defined on `Any`](https://www.scala-lang.org/api/2.13.1/scala/Any.html#!=(x$1:Any):Boolean), I think this solution makes the most sense. It also emphasizes that equality is between to `Any` values, and that `1 == "1"` will return `false`.

**Comment by frobinet** [comment]
This is a very interesting idea and would have many applications in anything related to numerics. From the looks of it, it could help with having compile-time shape checks for something like Tensorflow!

**Comment by MaximeKjaer** [comment]
@frobinet I haven't mentioned it in the PR, but that's actually exactly what I'm hoping to achieve with this feature! I've been working on that as my semester project at EPFL over the past few months, and I plan on open-sourcing my implementation within the next two weeks or so.

There are indeed many good applications for this feature. For instance, a paper called "[Dex: array programming with typed indices](https://openreview.net/pdf?id=rJxd7vsWPS)" mentions the following:

> Is our Hindley-Milner-based type language expressive enough to admit the programs that numerical programmers want to write? What about the dreaded `reshape`, which would seem to require type-level arithmetic? 

Well, that's what this PR implements :tada: It would make it possible to type operations like `reshape` in Scala.

**Comment by frobinet** [comment]
Sounds great. The only place I've seen this is C++ where you can just stick random computations inside template definitions, but that's just fancy copy-pasting. I guess it would be hard to get the reshape correct there, not to mention the horrendous error messages the compiler would throw at you if you get shapes wrong. Anyway, interesting project, I'll follow it one you're open-source!

File: library/src/scala/compiletime/ops/package.scala [hunk 9]
```diff
@@ -0,0 +1,174 @@
+package scala.compiletime
+
+import scala.annotation.infix
+
+package object ops {
+  object any {
+    /** Equality comparison of two singleton types.
+     * ```scala
```
Comment (OlivierBlanvillain) [inline library/src/scala/compiletime/ops/package.scala]: No need to specify scala here: that's the default inside scaladoc.

File: library/src/scala/compiletime/ops/package.scala [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    /** Concatenation of two `String` singleton types.
+     *  ```scala
+     *  val hello: "hello " + "world" = "hello world"
+     *  ```
+     */
+    @infix type +[X <: String, Y <: String] <: String
+  }
+
+  object int {
+    /** Addition of two `Int` singleton types. Type-level equivalent of `scala.Int.+`
```
Comment (OlivierBlanvillain) [inline library/src/scala/compiletime/ops/package.scala]: I would remove all the `Type-level equivalant` comments since it's pretty obvious. Also it's inaccurate as `+` is not actually a method on `object Int`...

Comment (MaximeKjaer) [reply library/src/scala/compiletime/ops/package.scala]: All right, fixed in 7e0a1db10f4a1e35872fa04c2426ede2b6a6a004.

File: library/src/scala/compiletime/ops/package.scala (hunk 9 above)
Comment (MaximeKjaer) [reply library/src/scala/compiletime/ops/package.scala]: This doesn't seem to be true. In the screenshot below, `&&` has the `scala` annotation on the Markdown code block, and `||` does not. I've therefore left the annotations in.

image
